### DISCUSSION TITLE
Sending from SX1276 to NRF905

### DISCUSSION THREAD

**Discussion by burk** [jgromes/RadioLib#978] [Q&A]
Hello!

I'm trying to use a TTGO LoRa32 SX1276 to communicate with some heaters I have which use NRF905 radio modules. The NRF905 communicates with:
 - 868.4Mhz
 - 100 kbits per second inverted Manchester encoding, so 50 kbits per second real data rate
 - 10 bit preamble `1111110101` which is also Manchester encoded into 20 bits
 - 4 byte address used as sync word (but I want to listen to all addresses)
 - payload and CRC at the end
 
I've managed to receive packets with the following code:
```
 uint8_t syncWord[] = {0x55, 0x99};
 int state = radio.beginFSK(
    868.4,  // Frequency,
    100.0,  // Bit rate (kbps).
    50.0,   // Frequency deviation.
    100.0,  // Receiver bandwidth.
    10,     // Output power (dbm).
    8,      // Preamble length in bits.
    false); // No OOK.

  if (state == RADIOLIB_ERR_NONE)
  {
    Serial.println(F("success!"));
    radio.packetMode();
    radio.setGain(0); // Automatic gain control.
    radio.setSyncWord(syncWord, 2);
    radio.setDataShaping(RADIOLIB_SHAPING_0_5);
    radio.setEncoding(RADIOLIB_ENCODING_MANCHESTER);
    radio.setCrcFiltering(false);
    radio.fixedPacketLengthMode(LENGTH);
    radio.setCRC(false);
    radio.setCurrentLimit(120);
  }
```
The sync word is set so that it maches 16 out of the 20 preamble bits of the NRF905. However I'm not sure why it works to set preamble bits to 8. Maybe it should rather be 0? But I got the impression from another thread that that is not supported.

I now want to send packages. I tried just doing `radio.transmit(...)` without success. Unfortunately I do not have control over the NRF905 side of the communication, so there could be something wrong with my payload.

But I also suspect there is something off with my SX1276 setup. Any advice? Is there a way to recreate the 10 bit (20 bit after Manchester) preamble of the NRF905? Do I need to switch over to the direct interface maybe?

**Sources:**
NRF905 parameters: https://github.com/mirakonta/NRF905-emulation-with-SX1276
Arduino NRF905 library and protocol: https://blog.zakkemble.net/nrf905-avrarduino-librarydriver/
Preamble description: https://devzone.nordicsemi.com/f/nordic-q-a/47122/nrf905-preamble

**Comment by burk** [comment]
This is really my first time doing anything with RF, so I'm fumbling a bit in the dark. As a debugging step I tried looking at the spectrum in URH (using an RTL-SDR dongle).

This is the spectrum when transmitting from the SX1276:
![Screenshot from 2024-02-18 12-56-48](https://github.com/jgromes/RadioLib/assets/1082217/2683e8e9-a95e-4f63-b0ea-0cc9bb5cf8cb)

This is the spectrum when the NRF905 is transmitting:
![Screenshot from 2024-02-18 12-57-33](https://github.com/jgromes/RadioLib/assets/1082217/7198de52-8e52-46ae-8f54-7a266430d79c)

(the NRF905 was further away, while the SX1276 was right beside the SDR dongle, if that matters)

**Comment by jgromes** [comment]
> However I'm not sure why it works to set preamble bits to 8. Maybe it should rather be 0? But I got the impression from another thread that that is not supported.

The library won't stop you from setting preamble to 0, though I'm not sure it will be able to receive anything at that point. However, since you have set your sync word to 0x5599, the first 8 preamble bits will be an additional 0x55, which matches the preamble sent by the nRF. Had you not done that, your preamble would be 4 bits too short.

Of course, when you're transmitting, it will be 4 bits too long. That could be one of the things that the nRF doesn't like. Unfortuantely SX127x can only set preamble length in whole bytes. You could try doing this manually, by shifting everything in the packet by 4 bits (that would mean the end of the RadioLib sync word would contain the 4 most significant bits of the payload) - might be worth a shot so that the nRF doesn't see any extra preamble bits which it might not be expecting. Other than that, I think you have everything correctly.

**Reply by burk** [reply]
Hmm that's a nice idea. I'll try that.

I'll can also try setting the preamble to 0, and sync word to `0x055599` which should match the nRF (if the SX1276 accepts the 0 length preamble setting).

**Comment by VirusPilot** [comment]
In a similar application (FLARM) we have been facing a similar situation (including also the SX1262), here is how it has been resolved:
- https://gist.github.com/VirusPilot/f8ca5548d87002fa10e2cd86502fec3a#flarm-rf-packet-nrf905-implementation
- https://gist.github.com/VirusPilot/f8ca5548d87002fa10e2cd86502fec3a#flarm-rf-packet-sx1262sx1276-implementation
